File: data-provider/src/main/scala/velocorner/feed/OpenWeatherFeed.scala
package velocorner.feed

import com.typesafe.scalalogging.LazyLogging
import velocorner.SecretConfig
import velocorner.model.weather.{ForecastResponse, WeatherResponse}
import velocorner.util.JsonIo
import cats.implicits._

import scala.concurrent.Future

object OpenWeatherFeed {

  val baseUrl = "https://api.openweathermap.org/data/2.5"
}

class OpenWeatherFeed(val config: SecretConfig) extends HttpFeed with LazyLogging {

  // for 5 days
  def forecast(location: String): Future[ForecastResponse] = {
    logger.debug(s"retrieving forecast for $location")
    val response = ws{_.url(s"${OpenWeatherFeed.baseUrl}/forecast")
      .withQueryStringParameters(
        ("q", location),
        ("appid", config.getId("weather")),
        ("units", "metric"),
        ("lang", "en")
      )
      .get()
    }
    response
      .map(_.body)
      .map(JsonIo.read[ForecastResponse])
  }

  def current(location: String): Future[Option[WeatherResponse]] = {
    logger.debug(s"retrieving current weather for $location")
    val response = ws{_.url(s"${OpenWeatherFeed.baseUrl}/weather")
      .withQueryStringParameters(
        ("q", location),
        ("appid", config.getId("weather")),
        ("units", "metric"),
        ("lang", "en")
      )
      .get()
    }
    response.map { res => res.status match {
      case 200 => JsonIo.read[WeatherResponse](res.body).some
      case _ => None
    }}

  }
}

File: src/main/kotlin/velocorner/weather/service/OpenWeatherFeed.kt
package velocorner.weather.service

import io.ktor.client.*
import io.ktor.client.request.*
import io.ktor.client.statement.*
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.json.Json
import velocorner.weather.model.CurrentWeatherResponse
import velocorner.weather.model.ForecastWeatherResponse

class OpenWeatherFeed(val apiKey: String, val client: HttpClient) {

    private val baseUrl = "https://api.openweathermap.org/data/2.5"
    private val json = Json {
        ignoreUnknownKeys = true
    }

    suspend fun current(location: String): CurrentWeatherResponse? =
        get<CurrentWeatherResponse>("weather", location)

    suspend fun forecast(location: String): ForecastWeatherResponse? =
        get<ForecastWeatherResponse>("forecast", location)

    internal suspend inline fun <reified T> get(path: String, location: String): T? {
        val response = client.get("$baseUrl/$path") {
            parameter("q", location)
            parameter("appid", apiKey)
            parameter("units", "metric")
            parameter("lang", "en")
        }
        return runCatching { json.decodeFromString<T>(response.bodyAsText()) }.getOrNull()
    }
}